x-pack/lib/helpers/elasticsearch_options.rb

```diff
@@ -27,53 +44,48 @@ def es_options_from_settings_or_modules(feature, settings)
-        opts['password'] = settings.get("#{prefix}#{feature}.elasticsearch.password")
-      end
-      if proxysetting = settings.get("#{prefix}#{feature}.elasticsearch.proxy")
-        opts['proxy'] = proxysetting
+      validate_authentication!(feature, settings, prefix)
+
+      # transpose all directly mappable settings
+      SETTINGS_MAPPINGS.each do |xpack_setting, es_setting|
+        v = settings.get("#{prefix}#{feature}.elasticsearch.#{xpack_setting}")
+        opts[es_setting] = v unless v.nil?
       end
 
-      opts['sniffing'] = settings.get("#{prefix}#{feature}.elasticsearch.sniffing")
-      opts['ssl_certificate_verification'] = settings.get("#{prefix}#{feature}.elasticsearch.ssl.verification_mode") == 'certificate'
+      # process remaining settings
 
-      if cacert = settings.get("#{prefix}#{feature}.elasticsearch.ssl.certificate_authority")
-        opts['cacert'] = cacert
-        opts['ssl'] = true
+      unless settings.get("#{prefix}#{feature}.elasticsearch.cloud_id")
+        opts['hosts'] = settings.get("#{prefix}#{feature}.elasticsearch.hosts")
       end
+      opts['ssl_certificate_verification'] = settings.get("#{prefix}#{feature}.elasticsearch.ssl.verification_mode") == 'certificate'
 
-      if truststore = settings.get("#{prefix}#{feature}.elasticsearch.ssl.truststore.path")
-        opts['truststore'] = truststore
-        opts['truststore_password'] = settings.get("#{prefix}#{feature}.elasticsearch.ssl.truststore.password")
+      # if all hosts are using https or any of the ssl related settings are set
+      if ssl?(feature, settings, prefix)
         opts['ssl'] = true
       end
 
-      if keystore = settings.get("#{prefix}#{feature}.elasticsearch.ssl.keystore.path")
-        opts['keystore'] = keystore
-        opts['keystore_password']= settings.get("#{prefix}#{feature}.elasticsearch.ssl.keystore.password")
-        opts['ssl'] = true
+      # the username setting has a default value and should not be included if using another authentication
+      if settings.set?("#{prefix}#{feature}.elasticsearch.cloud_auth") || settings.set?("#{prefix}#{feature}.elasticsearch.api_key")
+        opts.delete('user')
       end
+
       opts
     end
 
+    def ssl?(feature, settings, prefix)
+      verify_https_scheme(feature, settings, prefix) || settings.get("#{prefix}#{feature}.elasticsearch.ssl.certificate_authority") ||
+      (settings.get("#{prefix}#{feature}.elasticsearch.ssl.truststore.path") && settings.get("#{prefix}#{feature}.elasticsearch.ssl.truststore.password")) ||
+      (settings.get("#{prefix}#{feature}.elasticsearch.ssl.keystore.path") && settings.get("#{prefix}#{feature}.elasticsearch.ssl.keystore.password"))
+    end
+
+    HTTPS_SCHEME = /^\s*https:\/\/.+/
+    def verify_https_scheme(feature, settings, prefix)
+      hosts = Array(settings.get("#{prefix}#{feature}.elasticsearch.hosts"))
+      hosts.all? {|host| !host.match(HTTPS_SCHEME).nil?}
```

Comment:
I think `String#match?` may be a more readable fit here.

```suggestion
      hosts.all? {|host| host.match?(HTTPS_SCHEME)}
```